core/simulations/configs/vaultLiquidation.ts

```diff
@@ -0,0 +1,137 @@
+            choices: Object.keys(COLLATERALS)
+                .sort()
+                .map(collateral => ({
+                    title: collateral,
+                    value: collateral,
+                })),
+        },
+    ]);
+    return collateralType;
+};
+
+const getCollateralType = memoizee(_getCollateralType, {
+    promise: true,
+});
+
+const _getLatestVault = async () => {
+    const cdpManager = await getContract(TEST_NETWORK, 'CDP_MANAGER', true);
+    const lastHex = await cdpManager.last(HARDHAT_PUBLIC_KEY);
+    return new BigNumber(lastHex._hex).toNumber();
+};
+
+const getLatestVaultIndex = memoizee(_getLatestVault, { promise: true });
+
+const simulation: Simulation = {
+    title: 'Simulate liquidation Auctions',
+    steps: [
+        {
+            title: 'Reset blockchain fork',
+            // Few blocks before WSTETH-A is taken at 14052147,
+            // https://etherscan.io/address/0x49a33a28c4c7d9576ab28898f4c9ac7e52ea457at
+            entry: async () => {
+                await resetNetworkAndSetupWallet(14052140);
+            },
+        },
+        {
+            title: 'Set collateral in VAT',
+            entry: async () => {
+                const collateralType = await getCollateralType();
+                await setCollateralInVat(collateralType, COLLATARAL_OWNED);
+            },
+        },
+        {
+            title: 'Open the vault',
+            entry: async () => {
+                const collateralType = await getCollateralType();
+                await openVault(TEST_NETWORK, HARDHAT_PUBLIC_KEY, collateralType);
+            },
+        },
+        {
+            title: 'Extract collateral',
+            entry: async () => {
+                const collateralType = await getCollateralType();
+                const addressJoin = await getContractAddressByName(
+                    TEST_NETWORK,
+                    getJoinNameByCollateralType(collateralType)
+                );
+                const collateralConfig = getCollateralConfigByType(collateralType);
+                const tokenContractAddress = await getContractAddressByName(TEST_NETWORK, collateralConfig.symbol);
+                const contract = await getErc20Contract(TEST_NETWORK, tokenContractAddress, true);
+                await contract.approve(addressJoin, COLLATARAL_OWNED.shiftedBy(WAD_NUMBER_OF_DIGITS).toFixed(0));
+                await withdrawCollateralFromVat(TEST_NETWORK, HARDHAT_PUBLIC_KEY, collateralType, undefined);
+            },
+        },
+        {
+            title: 'Deposit Collateral to vault',
+            entry: async () => {
+                const latestVaultId = await getLatestVaultIndex();
+                const vault = await fetchVaultBase(TEST_NETWORK, latestVaultId);
+                await depositCollateralToVat(
+                    TEST_NETWORK,
+                    vault.address,
+                    vault.collateralType,
+                    COLLATARAL_OWNED.dividedBy(1.2)
+                );
```

Comment:
advice requested @valiafetisov 

This step fails with `GemJoin/failed-transfer` that corresponds to failed transfer in the [join.sol](https://github.com/makerdao/dss/blob/master/src/join.sol) the allowance has been given out in the previous step.

Vat operation (generating the collateral in vat) comes before transfer and hence does not fail.

I'm short on ideas what is going wrong here

Reply:
- i have this in the wallet
- __I__ have the amount in the VAT initially, but i do `join` so that it ends up in the __vault's__ VAT.
- i mean: possible to bypass this by adding the vat directly into the vault's VAT, but then it's not a 100% realistic simulation

Reply:
Sorry, I just skimmed it though last time and didn't get the whole flow, now I get it. My suggestion to move forward: to validate every step (that money is in the account, then that allowance was given, etc) and also try to authorize. After I given the authorization, I received a different type of error (but on the same simulation step):

<details><summary>SafeMath: subtraction overflow error</summary>

```
✔ Press Enter to proceed to the next step "Deposit Collateral to vault" … 
transaction 3d793d28-2a28-4281-8e7e-cc9c286acbf2: Transaction is preparing to be mined...
transaction 3d793d28-2a28-4281-8e7e-cc9c286acbf2: Transaction error: "Error: VM Exception while processing transaction: reverted with reason string SafeMath: subtraction overflow (cannot estimate gas; transaction may fail or may require manual gas limit [ See: https://links.ethers.org/v5-errors-UNPREDICTABLE_GAS_LIMIT ])"

/Users/user/Projects/sidestream/unified-auctions-ui/core/node_modules/@ethersproject/logger/src.ts/index.ts:261
        const error: any = new Error(message);
                           ^
Error: cannot estimate gas; transaction may fail or may require manual gas limit [ See: https://links.ethers.org/v5-errors-UNPREDICTABLE_GAS_LIMIT ] (error={"reason":"Error: VM Exception while processing transaction: reverted with reason string 'SafeMath: subtraction overflow'","code":"UNPREDICTABLE_GAS_LIMIT","method":"estimateGas","transaction":{"from":"0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266","maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x59682f00"},"maxFeePerGas":{"type":"BigNumber","hex":"0x59682f02"},"to":"0x24e459F61cEAa7b1cE70Dbaea938940A7c5aD46e","data":"0x3b4da69f000000000000000000000000bcca2b4c57ac039f14e2f125e3bfaffbcda6356600000000000000000000000000000004e5dc52f5e4c61ff1f453902aaaaaaaaa","type":2,"accessList":null},"error":{"reason":"processing response error","code":"SERVER_ERROR","body":"{\"jsonrpc\":\"2.0\",\"id\":60,\"error\":{\"code\":-32603,\"message\":\"Error: VM Exception while processing transaction: reverted with reason string 'SafeMath: subtraction overflow'\",\"data\":{\"message\":\"Error: VM Exception while processing transaction: reverted with reason string 'SafeMath: subtraction overflow'\",\"data\":\"0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001e536166654d6174683a207375627472616374696f6e206f766572666c6f770000\"}}}","error":{"code":-32603,"data":{"message":"Error: VM Exception while processing transaction: reverted with reason string 'SafeMath: subtraction overflow'","data":"0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001e536166654d6174683a207375627472616374696f6e206f766572666c6f770000"}},"requestBody":"{\"method\":\"eth_estimateGas\",\"params\":[{\"type\":\"0x2\",\"maxFeePerGas\":\"0x59682f02\",\"maxPriorityFeePerGas\":\"0x59682f00\",\"from\":\"0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266\",\"to\":\"0x24e459f61ceaa7b1ce70dbaea938940a7c5ad46e\",\"data\":\"0x3b4da69f000000000000000000000000bcca2b4c57ac039f14e2f125e3bfaffbcda6356600000000000000000000000000000004e5dc52f5e4c61ff1f453902aaaaaaaaa\"}],\"id\":60,\"jsonrpc\":\"2.0\"}","requestMethod":"POST","url":"http://127.0.0.1:8545"}}, tx={"data":"0x3b4da69f000000000000000000000000bcca2b4c57ac039f14e2f125e3bfaffbcda6356600000000000000000000000000000004e5dc52f5e4c61ff1f453902aaaaaaaaa","to":{},"maxFeePerGas":{"type":"BigNumber","hex":"0x59682f02"},"maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x59682f00"},"from":"0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266","type":2,"nonce":{},"gasLimit":{},"chainId":{}}, code=UNPREDICTABLE_GAS_LIMIT, version=abstract-signer/5.6.2)
    at Logger.makeError (/Users/user/Projects/sidestream/unified-auctions-ui/core/node_modules/@ethersproject/logger/src.ts/index.ts:261:28)
    at Logger.throwError (/Users/user/Projects/sidestream/unified-auctions-ui/core/node_modules/@ethersproject/logger/src.ts/index.ts:273:20)
    at /Users/user/Projects/sidestream/unified-auctions-ui/core/node_modules/@ethersproject/abstract-signer/src.ts/index.ts:301:31
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Promise.all (index 7)
```
</details>

Reply:
>  My suggestion to move forward: to validate every step

👍🏻 

> I given the authorization to this collateral

So, reading our code i see that we have 3 authorizations: 
  - to the clipper,
  - to the flapper,
  - to the flopper.

each of those is responsible to manage the auctions.

I have 2 questions:
  - which function did you call via ui? (i need it in order to find out which contract got your auth at the end)
  - did you guess that the auth is needed here or do you have a specific place to reference that shows that it's needed?

Reply:
> did you guess that the auth is needed here or do you have a specific place to reference that shows that it's needed?

I just called whatever auth we have on the frontend, I think it's `MCD_VAT.hope(clipperAddress)` aka `authorizeCollateral`. We require this authorisation to withdraw collateral from VAT in `Bidding with DAI` flow. Maybe you need to do it for the vault-related contracts instead? I can't see any auth in the code, but I had this in the past that it's hard to trace it.

I also suggested to fetch VAT balance, balance and allowance once you change them to validate they hold expected values, since it's nice to debug everything step-by-step and make sure that the current step doesn't fail because the previous didn't actually work

Reply:
In [dai.js source](https://github.com/makerdao/dai.js/blob/76f7ddc842d48e0eebdb4368f84e693509468cac/packages/dai-plugin-mcd/src/CdpManager.ts#L138-L179) they are calling quite a different method, if I understand correctly. If it's that complicated, we can directly modify vault amount if it doesn't break any further logic and let us liquidate the vault in the end

Reply:
Ok, turns out that indeed auth is not required here. Fixed the problem. The bad part was that i messed up the units. (WAD shifts)